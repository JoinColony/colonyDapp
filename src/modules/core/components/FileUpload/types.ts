import { FieldArrayRenderProps } from 'formik';

export interface FileReaderFile {
  name: string;
  type: string;
  size: number;
  lastModified: string;
  uploadDate: Date;
  data: string;
}

export interface UploadFile {
  file: File;
  uploaded?: string | boolean;
  error?: string;
  preview?: string;
  parsedData?: string[];
}

export interface UploadItemComponentProps {
  accept: string[];
  error?: string;
  key: string;
  idx: number;
  maxFileSize: number;
  name: string;
  remove: FieldArrayRenderProps['remove'];
  reset: FieldArrayRenderProps['form']['resetForm'];
  upload: UploadFn;
  validate: ValidateFileFn;
  handleError?: (...args: any[]) => Promise<any>;
  processingData?: boolean;
  handleProcessingData?: (...args: any) => void;
}

export type UploadFn = (fileData: FileReaderFile | File | null) => any;

export type ValidateFileFn = (value: UploadFile) => string | undefined;
