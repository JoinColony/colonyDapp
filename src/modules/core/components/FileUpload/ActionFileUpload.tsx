import React from 'react';

import { useAsyncFunction } from '~utils/hooks';

import FileUpload from './FileUpload';

import { ActionTransformFnType } from '~utils/actions';

interface Props {
  submit: string;
  success: string;
  error: string;
  transform?: ActionTransformFnType;
  accept?: string[];
  maxFileSize?: number;
  name?: string;
  label?: any;
  status?: any;
}

const ActionFileUpload = ({
  submit,
  success,
  error,
  transform,
  ...props
}: Props) => {
  const asyncFunction = useAsyncFunction({ submit, error, success, transform });
  return <FileUpload upload={asyncFunction} {...props} />;
};

export default ActionFileUpload;
