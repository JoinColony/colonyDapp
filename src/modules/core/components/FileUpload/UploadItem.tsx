import React, { SyntheticEvent, useEffect, useCallback, useMemo } from 'react';
import { defineMessages } from 'react-intl';

import { useField } from 'formik';
import { log } from '~utils/debug';

import { UploadFile, UploadItemComponentProps } from './types';
import Button from '../Button';
import Icon from '../Icon';
import { Tooltip } from '../Popover';
import ProgressBar from '../ProgressBar';

import fileReader from '../../../../lib/fileReader';

import styles from './UploadItem.css';

const MSG = defineMessages({
  removeActionText: {
    id: 'UploadItem.removeActionText',
    defaultMessage: 'Remove',
  },
});

const displayName = 'UploadItem';

const UploadItem = ({
  accept,
  error,
  idx,
  maxFileSize,
  name,
  remove,
  upload,
}: UploadItemComponentProps) => {
  const [
    ,
    {
      value: { file, uploaded },
      value,
    },
    { setValue },
  ] = useField<UploadFile>(name);

  // eslint-disable-next-line no-underscore-dangle
  const _readFiles = useMemo(
    () =>
      fileReader({
        maxFilesLimit: 1,
        maxFileSize,
        allowedTypes: accept,
      }),
    [accept, maxFileSize],
  );

  const read = useCallback(async () => {
    const [contents] = await _readFiles([file]);
    return contents;
  }, [_readFiles, file]);

  const handleRemoveClick = (evt: SyntheticEvent<HTMLButtonElement>) => {
    evt.stopPropagation();
    remove(idx);
  };

  const uploadFile = useCallback(async () => {
    if (!setValue) return;
    let readFile;
    let fileReference;
    try {
      readFile = await read();
      setValue({ ...value, preview: readFile.data });
      fileReference = await upload(
        value.file.type === 'text/csv' ? value.file : readFile,
      );
    } catch (caughtError) {
      log.error(caughtError);

      /**
       * @todo  better error handling here
       */
      setValue({ ...value, error: 'uploadError' });
      return;
    }
    setValue({ ...value, preview: readFile.data, uploaded: fileReference });
  }, [read, setValue, upload, value]);

  useEffect(() => {
    if (file && !error && !uploaded) {
      uploadFile();
    }
    // Only upload on mount
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div className={styles.uploadItem} aria-invalid={!!error}>
      <div className={styles.fileInfo}>
        <Tooltip
          placement="left"
          content={error || null}
          trigger={error ? 'hover' : 'disabled'}
        >
          <span className={styles.itemIcon}>
            <Icon name="file" title={file.name} />
          </span>
        </Tooltip>
        {uploaded || error ? (
          <span>{file.name}</span>
        ) : (
          <div className={styles.itemProgress}>
            <ProgressBar value={uploaded ? 100 : 0} />
          </div>
        )}
      </div>
      <div>
        <Button
          type="button"
          onClick={handleRemoveClick}
          appearance={{ theme: 'blue' }}
          text={MSG.removeActionText}
        />
      </div>
    </div>
  );
};

UploadItem.displayName = displayName;

export default UploadItem;
