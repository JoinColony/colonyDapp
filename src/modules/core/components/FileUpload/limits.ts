export const DEFAULT_MIME_TYPES: string[] = [
  'image/svg+xml',
  'image/png',
  'image/jpeg',
];

export const DEFAULT_MAX_FILE_SIZE = 1048576; // 1MB
