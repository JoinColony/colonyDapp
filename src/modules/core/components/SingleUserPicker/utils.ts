import { AnyUser } from '~data/index';

/*
Extracts the required values to be used in the SingleUserPicker
on selection
*/
export const filterUserSelection = (data: AnyUser[], filterValue: string) => {
  if (!filterValue) {
    return data;
  }

  const filteredUsers = data.filter((user) => {
    const username = user.profile?.username || '';
    const walletAddress = user.profile?.walletAddress || '';

    return (
      username.toLowerCase().includes(filterValue.toLowerCase()) ||
      walletAddress.toLowerCase().includes(filterValue.toLowerCase()) ||
      user.id.toLowerCase().includes(filterValue.toLowerCase())
    );
  });

  const customUserValue: AnyUser = {
    id: 'filterValue',
    profile: {
      walletAddress: filterValue,
      displayName: filterValue,
    },
  };

  return [customUserValue].concat(filteredUsers);
};
