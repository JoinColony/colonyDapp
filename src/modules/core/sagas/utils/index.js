/* @flow */

import getDDB from './getDDB';
import getNetworkClient from './getNetworkClient';

export * from './transactions';
export { getDDB, getNetworkClient };
