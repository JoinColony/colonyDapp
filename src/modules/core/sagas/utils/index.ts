export { default as getColonyManager } from './getColonyManager';
export { default as getGasPrices } from './getGasPrices';
export { getMethod, getTransactionMethod } from './getMethod';
export { default as getWalletCategory } from './getWalletCategory';
