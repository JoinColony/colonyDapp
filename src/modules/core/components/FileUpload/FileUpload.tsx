import React, {
  ComponentType,
  ReactNode,
  useCallback,
  useMemo,
  useImperativeHandle,
  Ref,
} from 'react';
import { MessageDescriptor, defineMessages } from 'react-intl';
import { DropzoneOptions, DropzoneState, useDropzone } from 'react-dropzone';
import { getIn } from 'formik';
import { compose } from 'recompose';

import { AsFieldArrayEnhancedProps } from '~core/Fields/asFieldArray';
import { Appearance } from '~core/Fields/Input/InputComponent';
import { SimpleMessageValues } from '~types/index';
import { withForwardingRef, ForwardedRefProps } from '~utils/hoc';

import DefaultPlaceholder from './DefaultPlaceholder';
import { DEFAULT_MIME_TYPES, DEFAULT_MAX_FILE_SIZE } from './limits';
import { UploadFn, UploadItemComponentProps, ValidateFileFn } from './types';
import UploadItem from './UploadItem';

import { asFieldArray } from '../Fields';
import InputLabel from '../Fields/InputLabel';
import InputStatus, { InputStatusAppearance } from '../Fields/InputStatus';

import styles from './FileUpload.css';

const MSG = defineMessages({
  labelError: {
    id: 'FileUpload.labelError',
    defaultMessage: 'There was an error processing your file. Try again.',
  },
  uploadError: {
    id: 'FileUpload.uploadError',
    defaultMessage: 'There was an error uploading your file',
  },
  filetypeError: {
    id: 'FileUpload.filetypeError',
    defaultMessage: 'This filetype is not allowed or file is too big',
  },
});

interface Props {
  /** Content to render inside the dropzone */
  children?: ReactNode | ((props: DropzoneState) => ReactNode);
  /** Custom classNames for different elements of the component */
  classNames?: {
    main?: string;
    dropzone?: string;
    dropzoneAccept?: string;
    dropzoneReject?: string;
    filesContainer?: string;
    disabled?: string;
  };
  /** Options for the dropzone provider */
  dropzoneOptions?: DropzoneOptions;
  /** Whether to display the element all by itself, or with labels & whatnot */
  elementOnly?: boolean;
  /** Extra node to render on the top right in the label */
  extra?: ReactNode;
  /** Help text (will appear next to label text) */
  help?: string | MessageDescriptor;
  /** Values for help text (react-intl interpolation) */
  helpValues?: SimpleMessageValues;
  /** Element html `id` */
  id?: string;
  /** The component to render each item to be uploaded */
  itemComponent?: ComponentType<UploadItemComponentProps>;
  /** Label text */
  label?: string | MessageDescriptor;
  /** Values for label text (react-intl interpolation) */
  labelValues?: SimpleMessageValues;
  /** Maximum number of files to accept */
  maxFilesLimit?: number;
  /** Max file size */
  maxSize?: number;
  /** Input control `name` attribute */
  name: string;
  /** Placeholder element for when no files have been picked yet (renderProp) */
  renderPlaceholder?: ReactNode | null;
  /** Ref for programmatic opening */
  ref?: Ref<any>;
  /** Status text */
  status?: string | MessageDescriptor;
  /** Function to handle the actual uploading of the file */
  upload: UploadFn;
  /** Function to handle an upload error from the outside */
  handleError?: (...args: any[]) => Promise<any>;

  labelAppearance?: Appearance;

  customErrorMessage?: string | MessageDescriptor;

  inputStatusAppearance?: InputStatusAppearance;

  processingData?: boolean;

  handleProcessingData?: (...args: any) => void;
}

const validateFile: ValidateFileFn = (value) =>
  value.error ? MSG[value.error] : undefined;

const FileUpload = ({
  children,
  classNames = styles,
  dropzoneOptions: { accept: acceptProp, disabled, ...dropzoneOptions } = {},
  elementOnly,
  extra,
  form: { errors, resetForm, values },
  forwardedRef: ref,
  help,
  helpValues,
  id,
  itemComponent: FileUploaderItem = UploadItem,
  label,
  labelValues,
  maxFilesLimit = 1,
  maxSize = DEFAULT_MAX_FILE_SIZE,
  name,
  push,
  remove,
  renderPlaceholder = <DefaultPlaceholder />,
  status,
  upload,
  handleError,
  labelAppearance,
  customErrorMessage,
  inputStatusAppearance,
  processingData,
  handleProcessingData,
}: AsFieldArrayEnhancedProps<Props> & ForwardedRefProps) => {
  const files = useMemo(() => getIn(values, name) || [], [name, values]);
  const fileErrors = useMemo(() => getIn(errors, name) || [], [errors, name]);

  const maxFileLimitNotMet = files.length < maxFilesLimit;
  const hasError = !!fileErrors.length;

  const accept = useMemo<string[]>(() => {
    if (!acceptProp) {
      return DEFAULT_MIME_TYPES;
    }
    if (!Array.isArray(acceptProp)) {
      return [acceptProp];
    }
    return acceptProp;
  }, [acceptProp]);

  const onDropAccepted = useCallback(
    (acceptedFiles) => {
      const countAcceptedFiles = files.filter(({ error }) => !error);
      const newFiles = Array.from(acceptedFiles).slice(
        0,
        maxFilesLimit - countAcceptedFiles.length,
      );
      newFiles.forEach((file) => push({ file }));
    },
    [files, maxFilesLimit, push],
  );

  const onDropRejected = useCallback(
    (rejectedFiles) => {
      rejectedFiles
        .slice(0, maxFilesLimit - files.length)
        .forEach((file) => push({ ...file, error: 'filetypeError' }));
    },
    [files.length, maxFilesLimit, push],
  );

  const dropzoneState = useDropzone({
    accept,
    disabled: disabled || !maxFileLimitNotMet,
    maxSize,
    onDropAccepted,
    onDropRejected,
    // We can override the above properties by providing these in the `dropzoneOptions` prop
    ...dropzoneOptions,
  });

  const {
    getRootProps,
    getInputProps,
    isDragAccept,
    isDragReject,
    open,
  } = dropzoneState;

  const renderExtraChildren = useCallback(() => {
    if (!children) return null;
    if (typeof children === 'function') {
      return children(dropzoneState);
    }
    return children;
  }, [children, dropzoneState]);

  const dropzoneClassName = useMemo(() => {
    const classes = [
      classNames.dropzone,
      ...(disabled ? [classNames.disabled] : []),
    ];
    if (isDragAccept) {
      classes.push(classNames.dropzoneAccept);
    } else if (isDragReject) {
      classes.push(classNames.dropzoneReject);
    }
    return classes.join(' ');
  }, [
    classNames.dropzone,
    classNames.dropzoneAccept,
    classNames.dropzoneReject,
    classNames.disabled,
    disabled,
    isDragAccept,
    isDragReject,
  ]);

  useImperativeHandle(ref, () => ({
    open,
  }));

  return (
    <div className={classNames.main} id={id}>
      {!elementOnly && label && (
        <InputLabel
          appearance={labelAppearance}
          label={label}
          help={help}
          labelValues={labelValues}
          helpValues={helpValues}
          extra={extra}
        />
      )}
      <div
        className={dropzoneClassName}
        {...getRootProps()}
        data-test="avatarUpload"
      >
        <input {...getInputProps()} />
        {maxFileLimitNotMet && renderPlaceholder}
        {files && files.length > 0 && (
          <div className={classNames.filesContainer}>
            {files.map((file, idx) => (
              <FileUploaderItem
                accept={accept}
                error={file.error}
                key={`${file.name}-${file.size}`}
                idx={idx}
                maxFileSize={maxSize}
                name={`${name}.${idx}`}
                remove={remove}
                reset={resetForm}
                upload={upload}
                validate={validateFile}
                handleError={handleError}
                processingData={processingData}
                handleProcessingData={handleProcessingData}
              />
            ))}
          </div>
        )}
        {renderExtraChildren()}
      </div>
      <InputStatus
        appearance={inputStatusAppearance}
        status={status}
        error={hasError ? customErrorMessage || MSG.labelError : ''}
      />
    </div>
  );
};

const enhance = compose<
  AsFieldArrayEnhancedProps<Props> & ForwardedRefProps,
  Props
>(
  // @NOTE ordering matters here - refs first!
  withForwardingRef,
  asFieldArray(),
);

export default enhance(FileUpload) as ComponentType<Props>;
