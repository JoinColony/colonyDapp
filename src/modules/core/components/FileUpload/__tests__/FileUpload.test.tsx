/* eslint-env jest */

import React from 'react';
import { Formik } from 'formik';
import toJson from 'enzyme-to-json';

import { mountWithIntl } from '../../../../../__tests__/utils';

import FileUpload from '../FileUpload';

describe('FileUpload component', () => {
  test('Renders initial component', () => {
    const wrapper = mountWithIntl(
      <Formik initialValues={{}} onSubmit={() => undefined}>
        <FileUpload
          name="fileUpload"
          label="Basic file upload"
          help="Some help text"
          maxFilesLimit={3}
        />
      </Formik>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
