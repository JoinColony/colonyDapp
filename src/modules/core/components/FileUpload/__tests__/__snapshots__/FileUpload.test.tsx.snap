// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload component Renders initial component 1`] = `
<Formik
  initialValues={Object {}}
  onSubmit={[Function]}
>
  <Component
    help="Some help text"
    label="Basic file upload"
    maxFilesLimit={3}
    name="fileUpload"
  >
    <FormikConnect(FieldArrayInner)
      name="fileUpload"
      render={[Function]}
    >
      <FieldArrayInner
        formik={
          Object {
            "dirty": false,
            "errors": Object {},
            "getFieldHelpers": [Function],
            "getFieldMeta": [Function],
            "getFieldProps": [Function],
            "handleBlur": [Function],
            "handleChange": [Function],
            "handleReset": [Function],
            "handleSubmit": [Function],
            "initialErrors": Object {},
            "initialStatus": undefined,
            "initialTouched": Object {},
            "initialValues": Object {},
            "isSubmitting": false,
            "isValid": true,
            "isValidating": false,
            "registerField": [Function],
            "resetForm": [Function],
            "setErrors": [Function],
            "setFieldError": [Function],
            "setFieldTouched": [Function],
            "setFieldValue": [Function],
            "setFormikState": [Function],
            "setStatus": [Function],
            "setSubmitting": [Function],
            "setTouched": [Function],
            "setValues": [Function],
            "status": undefined,
            "submitCount": 0,
            "submitForm": [Function],
            "touched": Object {},
            "unregisterField": [Function],
            "validateField": [Function],
            "validateForm": [Function],
            "validateOnBlur": true,
            "validateOnChange": true,
            "validateOnMount": false,
            "values": Object {},
          }
        }
        name="fileUpload"
        render={[Function]}
        validateOnChange={true}
      >
        <FileUpload
          accept={
            Array [
              "image/svg+xml",
              "image/png",
              "image/jpeg",
            ]
          }
          classNames={Object {}}
          disableClick={false}
          form={
            Object {
              "dirty": false,
              "errors": Object {},
              "getFieldHelpers": [Function],
              "getFieldMeta": [Function],
              "getFieldProps": [Function],
              "handleBlur": [Function],
              "handleChange": [Function],
              "handleReset": [Function],
              "handleSubmit": [Function],
              "initialErrors": Object {},
              "initialStatus": undefined,
              "initialTouched": Object {},
              "initialValues": Object {},
              "isSubmitting": false,
              "isValid": true,
              "isValidating": false,
              "registerField": [Function],
              "resetForm": [Function],
              "setErrors": [Function],
              "setFieldError": [Function],
              "setFieldTouched": [Function],
              "setFieldValue": [Function],
              "setFormikState": [Function],
              "setStatus": [Function],
              "setSubmitting": [Function],
              "setTouched": [Function],
              "setValues": [Function],
              "status": undefined,
              "submitCount": 0,
              "submitForm": [Function],
              "touched": Object {},
              "unregisterField": [Function],
              "validateField": [Function],
              "validateForm": [Function],
              "validateOnBlur": true,
              "validateOnChange": true,
              "validateOnMount": false,
              "values": Object {},
            }
          }
          handleInsert={[Function]}
          handleMove={[Function]}
          handlePop={[Function]}
          handlePush={[Function]}
          handleRemove={[Function]}
          handleReplace={[Function]}
          handleSwap={[Function]}
          handleUnshift={[Function]}
          help="Some help text"
          insert={[Function]}
          itemComponent={[Function]}
          label="Basic file upload"
          maxFileSize={1048576}
          maxFilesLimit={3}
          move={[Function]}
          name="fileUpload"
          pop={[Function]}
          push={[Function]}
          remove={[Function]}
          renderPlaceholder={<Placeholder />}
          replace={[Function]}
          swap={[Function]}
          unshift={[Function]}
        >
          <div>
            <InputLabel
              help="Some help text"
              label="Basic file upload"
            >
              <label
                className=""
              >
                <span>
                  Basic file upload
                </span>
                <span>
                  Some help text
                </span>
              </label>
            </InputLabel>
            <n
              accept={
                Array [
                  "image/svg+xml",
                  "image/png",
                  "image/jpeg",
                ]
              }
              aria-invalid={false}
              data-test="avatarUploaderDrop"
              disableClick={false}
              disabled={false}
              getDataTransferItems={[Function]}
              inputProps={Object {}}
              maxSize={1048576}
              minSize={0}
              multiple={true}
              onDropAccepted={[Function]}
              onDropRejected={[Function]}
              preventDropOnDocument={true}
            >
              <div
                aria-disabled={false}
                aria-invalid={false}
                className=""
                data-test="avatarUploaderDrop"
                onClick={[Function]}
                onDragEnter={[Function]}
                onDragLeave={[Function]}
                onDragOver={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                style={
                  Object {
                    "borderColor": "#666",
                    "borderRadius": 5,
                    "borderStyle": "dashed",
                    "borderWidth": 2,
                    "height": 200,
                    "position": "relative",
                    "width": 200,
                  }
                }
              >
                <Placeholder>
                  <div>
                    <FormattedMessage
                      defaultMessage="Drag or {browse}"
                      id="FileUpload.dropzoneText"
                      values={
                        Object {
                          "browse": <span>
                            <FormattedMessage
                              defaultMessage="browse"
                              id="FileUpload.dropzoneTextBrowseAction"
                              values={Object {}}
                            />
                          </span>,
                        }
                      }
                    >
                      Drag or 
                      <span>
                        <FormattedMessage
                          defaultMessage="browse"
                          id="FileUpload.dropzoneTextBrowseAction"
                          values={Object {}}
                        >
                          browse
                        </FormattedMessage>
                      </span>
                    </FormattedMessage>
                  </div>
                </Placeholder>
                <input
                  accept={
                    Array [
                      "image/svg+xml",
                      "image/png",
                      "image/jpeg",
                    ]
                  }
                  autoComplete="off"
                  disabled={false}
                  multiple={true}
                  onChange={[Function]}
                  style={
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "opacity": 0.00001,
                      "pointerEvents": "none",
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                  type="file"
                />
              </div>
            </n>
            <InputStatus
              error=""
            >
              <p
                className=""
              />
            </InputStatus>
          </div>
        </FileUpload>
      </FieldArrayInner>
    </FormikConnect(FieldArrayInner)>
  </Component>
</Formik>
`;
